src/LaunchDarkly.ServerSdk/InMemoryFeatureStore.cs

```diff
@@ -15,7 +15,6 @@ namespace LaunchDarkly.Client
     public class InMemoryFeatureStore : IFeatureStore

```

Comment:
I also strongly suggest that you use ImmutableDictionary instead of Dictionary

Reply:
Writes to this data set are always going to be very infrequent compared to reads. If you're updating flags often enough that the duration of a dictionary update is interfering noticeably with flag retrieval, I'd say something is wrong.

That said, I believe the main reason we haven't used the immutable collections so far was simply to reduce the number of NuGet dependencies. I think on balance the value of using them probably outweighs that.

I wasn't 100% sure at first that your proposed solution really eliminates the need for a lock, but .NET does guarantee atomicity of updating an object, reference, right?

Reply:
Hmm, well, I think there is still a race condition with that. If you have two threads trying to write an update at roughly the same time to two different flags A & B, one will produce a new ImmutableDictionary that has only the A update, and the other will produce a new ImmutableDictionary that has only the B update - so there is no valid data set that you can simply swap into place _if_ you're not using a lock at all. You would just need to have a lock that's only used in the update path. Maybe that is what you had in mind anyway.

(In practice, the SDK never updates the data store from more than one thread, but we shouldn't rely on that)

Reply:
In any case the current fix is definitely appropriate.

Reply:
You have to lock the writers, so that only one writer can see the existing state, update, and assign.  Readers just need to get the latest pointer, which you should still ideally wrap in a lock, but that would be quick.

Reply:
Here is what I have to far for the immutable version.  Just need to add tests.  But FYI that I'm working on it.

https://github.com/JeffAshton/.net-client/commit/3b4249bad3e6fa17c38d9651f037e7b4b2e923cc